### PR TITLE
Refactor to allow being imported as module

### PR DESCRIPTION
**Pull request by jesusfcr**

### CODE REVIEW

**Review by manelmontilla — CHANGES_REQUESTED** [review]

File: backend/docker/docker.go [hunk 1]
```diff
@@ -38,57 +43,130 @@ type Client interface {
 	Pull(ctx context.Context, imageRef string) error
 }
 
+type RunConfig struct {
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go:47]: Done in d721507

File: backend/docker/docker.go [hunk 2]
```diff
@@ -38,57 +43,130 @@ type Client interface {
...
+		}
+	}
+
+	return "", errors.New("failed to determine Docker agent IP address")
 }
 
 // New created a new Docker backend using the given config, agent api addres and
 // CheckVars.
-func New(log log.Logger, cfg config.RegistryConfig, retryer Retryer, agentAddr string, vars backend.CheckVars) (*Docker, error) {
+func BuildBackend(log log.Logger, cfg config.Config, hook Hook) (backend.Backend, error) {
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: Done in 754cfb1

File: backend/docker/docker.go [hunk 3]
```diff
@@ -38,57 +43,130 @@ type Client interface {
...
+	ContainerStartOptions types.ContainerStartOptions
+}
+
 // Retryer represents the functions used by the docker backend for retrying
 // docker registry operations.
 type Retryer interface {
 	WithRetries(op string, exec func() error) error
 }
 
+// Hook allows to update the docker configuration just before the container creation
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: done in 185d5f1

File: backend/docker/docker.go [hunk 4]
```diff
@@ -109,7 +187,17 @@ func (b *Docker) Run(ctx context.Context, params backend.RunParams) (<-chan back
...
-	contID, err := b.cli.Create(ctx, cfg, "")
+
+	if b.hook != nil {
+		err := b.hook(params, &cfg)
+		if err != nil {
+			res <- backend.RunResult{Error: err}
+			return
+		}
+	}
+	// TODO: Use a propper Arch
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: It was related to the Platform parameter, but it's optional so I removed the TODO in 83b4ab2

File: backend/docker/docker.go [hunk 5]
```diff
@@ -130,9 +218,17 @@ func (b *Docker) run(ctx context.Context, params backend.RunParams, res chan<- b
...
 
+	resultC, errC := b.cli.ContainerWait(context.Background(), cc.ID, "")
 	var exit int64
-	exit, err = b.cli.ContainerWait(ctx, contID)
-	b.log.Debugf("container with ID %s finished with exit code %d", contID, exit)
+	select {
+	case err = <-errC:
+	case result := <-resultC:
+		if result.Error != nil {
+			// TODO: Review how to manage this, ... generate an error?
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: result.Error is not an error. Instead I create an error with result.Error.Message in 1fb2a80 and fix the ctx.

File: backend/docker/docker.go [hunk 6]
```diff
@@ -22,14 +25,16 @@ import (
...
 	"github.com/adevinta/vulcan-agent/log"
+	"github.com/adevinta/vulcan-agent/retryer"
 )
 
 const abortTimeout = 5 * time.Second
+const defaultDockerIfaceName = "docker0"
 
 // Client defines the shape of the docker client component needed by the docker
 // backend in order to be able to run checks.
 type Client interface {
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: Right, removed in 735987b

File: backend/docker/docker.go [hunk 7]
```diff
@@ -38,57 +43,130 @@ type Client interface {
...
 type Docker struct {
 	config    config.RegistryConfig
 	agentAddr string
 	checkVars backend.CheckVars
 	log       log.Logger
-	cli       Client //DockerClient
+	cli       *client.Client
+	auth      types.AuthConfig
 	retryer   Retryer
+	hook      Hook
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: What about updater / ConfigUpdater? See 40ad04b

File: backend/docker/docker.go [hunk 8]
```diff
@@ -38,57 +43,130 @@ type Client interface {
...
+}
+
 // Retryer represents the functions used by the docker backend for retrying
 // docker registry operations.
 type Retryer interface {
 	WithRetries(op string, exec func() error) error
 }
 
+// Hook allows to update the docker configuration just before the container creation
+type Hook func(backend.RunParams, *RunConfig) error
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: What about ConfigUpdater? See 40ad04b

File: agent/agent.go [hunk 9]
```diff
@@ -32,38 +31,11 @@ import (
...
 )
 
-// BackendCreator defines the shape of the function that will be called by the
-// function MainWithExitCode in order to create the backend that will run the
-// checks.
-type BackendCreator func(log.Logger, config.Config, backend.CheckVars) (backend.Backend, error)
-
-// MainWithExitCode executes the agent with the backend created by calling the
+// Run executes the agent with the backend created by calling the
 // passed BackendCreator. When the function finishes it returns an exit code of
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply agent/agent.go]: Is this enough 8fa102b ?

File: backend/docker/docker.go [hunk 10]
```diff
@@ -38,57 +43,130 @@ type Client interface {
...
+			connAddr.IP = ip
+			return connAddr.String(), nil
+		}
+	}
+
+	return "", errors.New("failed to determine Docker agent IP address")
 }
 
 // New created a new Docker backend using the given config, agent api addres and
 // CheckVars.
```
[the remark this answers is not in the document]
Comment (jesusfcr) [reply backend/docker/docker.go]: See 40ad04b

File: agent/agent.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (manelmontilla) [reply agent/agent.go]: It is, thanks!!

File: backend/docker/docker.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (manelmontilla) [reply backend/docker/docker.go]: Yep I like it, much better than my suggestion.

File: backend/docker/docker.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (manelmontilla) [reply backend/docker/docker.go]: True!!, I like the solution 👍

File: backend/docker/docker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (manelmontilla) [reply backend/docker/docker.go]: ok thanks.

File: backend/docker/docker.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (manelmontilla) [reply backend/docker/docker.go]: Yep, good name.

**Review by manelmontilla — APPROVED** [review]
LGTM

**Pull request merged**